### PR TITLE
[TA4197] fix(replica): preload issue due to large no of extents

### PR DESCRIPTION
**Pull request by utkarshmani1997**
Recently we have seen that if volume size is large then preload
takes time and hence openReplica fails due to timeouts and hence
neither of them succedes.

There are five calls where Preload is required
- When replica is in Initial state
- Replica received `open` signal from controller
- Replica is closed
- Replica is reloading
- Revert to snapshot

To solve the issue, we have removed it from the construct function
rather we are calling it after the sync is done.

Feature : Added `PreloadStatus` field in Replica structure, so status can be queried via REST api
There are three status:
```
Error    : In case of replica disconnection or syscall failure
Started  : If Preload is happening in the background
Done     : If Preload is completed. 
```
NOTE:
- Since we are running IO's and preload parallely, we may miss some block counts for iostats, this can be taken as improvement for the next iteration

Signed-off-by: Utkarsh Mani Tripathi <utkarsh.tripathi@mayadata.io>

### CODE REVIEW

**Comment by utkarshmani1997** [comment]
@vishnuitta Generating extents in travis in very short time is not possible. I have already added a test case in litmus to verify this. https://github.com/openebs/litmus/pull/384

**Comment by vishnuitta** [comment]
> @vishnuitta Generating extents in travis in very short time is not possible. I have already added a test case in litmus to verify this. [openebs/litmus#384](https://github.com/openebs/litmus/pull/384)

lets add automated tests to hit our modified code.. not need to get lot of extents

File: sync/sync.go [hunk 1]
```diff
@@ -358,8 +368,10 @@ Register:
 	}
 
 	logrus.Infof("reloadAndVerify %v", replicaAddress)
-	return t.reloadAndVerify(replicaAddress, toClient)
-
+	if err := t.reloadAndVerify(replicaAddress, toClient); err != nil {
+		return err
+	}
+	return s.Preload()
```
[the remark this answers is not in the document]
Comment (utkarshmani1997) [reply sync/sync.go]: Done

File: sync/sync.go [hunk 2]
```diff
@@ -358,8 +362,24 @@ Register:
...
+		return err
+	}
 
+	t.preload(s)
+	return nil
+}
+
+func (t *Task) preload(s *replica.Server) {
+	logrus.Info("Start reading extents")
+	go func() {
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply sync/sync.go]: good that with controller restart testcase, we found this issue of monitoring disconnection.
Regarding closing of go-routine, lock will just safeguard not to enter, but, better would be reuse of same go routine without creating new one (or) to kill the go-routine.

File: sync/sync.go [hunk 3]
```diff
@@ -358,8 +362,28 @@ Register:
...
+}
+
+func (t *Task) preload(s *replica.Server) {
+	logrus.Info("Start reading extents")
+	if s.PreloadStatus == "Started" {
+		logrus.Warning("Already reading extents in background")
+		return
+	}
+	go func() {
+		if err := s.Preload(); err != nil {
```
Comment (vishnuitta) [inline sync/sync.go]: have a testcase for error in preload

File: sync/sync.go [hunk 4]
```diff
@@ -358,8 +362,28 @@ Register:
...
+		return err
+	}
 
+	t.preload(s)
+	return nil
+}
+
+func (t *Task) preload(s *replica.Server) {
+	logrus.Info("Start reading extents")
+	if s.PreloadStatus == "Started" {
```
Comment (vishnuitta) [inline sync/sync.go]: looks good.. will review it keenly

File: replica/backup.go [hunk 5]
```diff
@@ -270,8 +272,12 @@ func preload(d *diffDisk) error {
 					length++
 				}
 			}
-			d.location[offset] = byte(i)
-			d.UsedBlocks++
+			d.rmLock.Lock()
+			if i != len(d.files) {
```
Comment (vishnuitta) [inline replica/backup.go]: shouldn't this be `d.location[offset] != len(d.files)`? and also, this assumes that new file is created after restart by the time preload starts. We need to handle that assumption.

File: replica/diff_disk.go [hunk 6]
```diff
@@ -19,6 +19,7 @@ type diffDisk struct {
 	// mapping of sector to index in the files array. a value of 0 is special meaning
 	// we don't know the location yet.
 	location          []byte
+	preloadStatus     string
```
Comment (vishnuitta) [inline replica/diff_disk.go]: make this as enum

Comment (vishnuitta) [reply replica/diff_disk.go]: also, make it visible in some rest api

File: replica/diff_disk.go [hunk 7]
```diff
@@ -311,7 +312,7 @@ func (d *diffDisk) lookup(sector int64) (byte, error) {
 	}
 
 	target := d.location[sector]
-	if target == 0 {
+	if target == 0 || d.preloadStatus == "Started" || d.preloadStatus == "Error" {
```
Comment (vishnuitta) [inline replica/diff_disk.go]: better would be d.prelodStatus != "Done"

File: replica/diff_disk.go (hunk 6 above)
Comment (vishnuitta) [reply replica/diff_disk.go]: also comment explaining each state

File: replica/server.go [hunk 8]
```diff
@@ -102,6 +104,20 @@ func (s *Server) Create(size int64) error {
...
+	if s.r == nil {
+		return fmt.Errorf("Can't read extents, s.r is nil")
+	}
+
+	s.PreloadStatus = "Started"
+	inject.AddPreloadTimeout()
+	if err := s.r.Preload(); err != nil {
+		return err
+	}
+	s.PreloadStatus = ""
```
Comment (vishnuitta) [inline replica/server.go]: why do we need to nullify this status?

File: replica/server.go [hunk 9]
```diff
@@ -102,6 +104,20 @@ func (s *Server) Create(size int64) error {
 	return r.Close()
 }
 
+func (s *Server) Preload() error {
+	if s.r == nil {
+		return fmt.Errorf("Can't read extents, s.r is nil")
+	}
+
+	s.PreloadStatus = "Started"
```
Comment (vishnuitta) [inline replica/server.go]: To avoid multiple of these go routines to get started, PreloadStatus need to be set to 'Started' even before this go routine is created. In case if you want to know that go-routine is started, you can have some other state.

File: sync/sync.go (hunk 4 above)
Comment (vishnuitta) [reply sync/sync.go]: As mentioned in one of above comment, PreloadStatus need to be set to 'Started' right after this check and before go routine is started.

File: replica/backup.go (hunk 5 above)
Comment (utkarshmani1997) [reply replica/backup.go]: I was ignoring the last file completely, although checking at particular offset is better.

File: replica/diff_disk.go (hunk 6 above)
Comment (utkarshmani1997) [reply replica/diff_disk.go]: done

File: replica/diff_disk.go (hunk 7 above)
Comment (utkarshmani1997) [reply replica/diff_disk.go]: ok

File: sync/sync.go (hunk 4 above)
Comment (utkarshmani1997) [reply sync/sync.go]: Sounds good

File: replica/backup.go [hunk 10]
```diff
@@ -270,8 +275,12 @@ func preload(d *diffDisk) error {
 					length++
 				}
 			}
-			d.location[offset] = byte(i)
-			d.UsedBlocks++
+			d.rmLock.Lock()
+			if d.location[offset] < byte(len(d.files)) {
```
Comment (vishnuitta) [inline replica/backup.go]: is this supposed to be d.location[offset] < (byte(len(d.files)) -1 )

Comment (utkarshmani1997) [reply replica/backup.go]: d.location[offset] != (byte(len(d.files))) is correct validation, explanation is as follows:
- d.files[0] = <nil>
- d.files[1]  = file1 
and so on
on each restart a new file get appended to d.files so the latest file will be at the last location. For example if there was 5 files initially and restart happens so total file is 5+1(new file) = 6 and size of d.files would be 6 + 1(<nil> oth location) = 7. So we will verify that if offset is in latest file or not, if it's already  there then we will ignore and if it's not there will update it.

File: replica/diff_disk.go [hunk 11]
```diff
@@ -242,6 +242,8 @@ func (d *diffDisk) fullReadAt(buf []byte, offset int64) (int, error) {
 		c   int
 	)
 
+	d.rmLock.Lock()
```
Comment (vishnuitta) [inline replica/diff_disk.go]: this can be very long lock. Why do we need this?

File: sync/sync.go [hunk 12]
```diff
@@ -381,8 +381,7 @@ func (t *Task) preload(s *replica.Server) {
 	go func() {
 		if err := s.Preload(); err != nil {
 			s.PreloadStatus = types.Error
-			logrus.Error(err)
-			s.MonitorChannel <- struct{}{}
+			logrus.Fatal("Exiting, err: ", err)
```
Comment (vishnuitta) [inline sync/sync.go]: have more explanatory log here

File: replica/server.go [hunk 13]
```diff
@@ -376,6 +395,18 @@ func (s *Server) Close(signalMonitor bool) error {
 		return nil
 	}
 
+	if s.PreloadStatus != types.Done {
```
Comment (vishnuitta) [inline replica/server.go]: why do we need a 'if' and then 'for' when you are however doing a 'log' in both the cases?

File: replica/backup.go [hunk 14]
```diff
@@ -270,8 +275,12 @@ func preload(d *diffDisk) error {
 					length++
 				}
 			}
-			d.location[offset] = byte(i)
-			d.UsedBlocks++
+			d.rmLock.Lock()
+			if d.location[offset] < byte(i) {
+				d.location[offset] = byte(i)
+				d.UsedBlocks++
```
Comment (vishnuitta) [inline replica/backup.go]: is UsedBlocks coming correct now even with this change?

File: sync/sync.go (hunk 12 above)
Comment (utkarshmani1997) [reply sync/sync.go]: err itself will be self explainatory.

File: replica/server.go (hunk 13 above)
Comment (utkarshmani1997) [reply replica/server.go]: I had used it for debugging purpose, removing it.

File: replica/backup.go (hunk 14 above)
Comment (utkarshmani1997) [reply replica/backup.go]: We are not using `usedblock` in ci. `du -sh` is used to verify the sizes.

Comment (vishnuitta) [reply replica/backup.go]: ok.. so, to fix this still, we can do like
```
if d.location[offset] != byte(i) {
        d.UsedBlocks++
}
```
And, is there a way to know the value of this d.UsedBlocks during runtime?

File: replica/diff_disk.go [hunk 15]
```diff
@@ -186,7 +183,9 @@ func (d *diffDisk) fullWriteAt(buf []byte, offset int64) (int, error) {
 				length++
 			}
 		}
+		d.rmLock.Lock()
```
Comment (vishnuitta) [inline replica/diff_disk.go]: `UsedBlocks` seems to be going wrong due to this and previous 'space reclaimation' PR.
Can you explain me the usage of `d.UsedBlocks`?

File: replica/diff_disk.go [hunk 16]
```diff
@@ -166,7 +163,7 @@ func (d *diffDisk) fullWriteAt(buf []byte, offset int64) (int, error) {
 		if val := d.location[startSector+i]; val == 0 {
 			d.UsedLogicalBlocks++
 			d.UsedBlocks++
-		} else if val != target {
+		} else if val != target && d.preloadStatus == types.Done {
```
Comment (vishnuitta) [inline replica/diff_disk.go]: you need to add a comment here, and also, at the end of the 'for' loop where sendToCreateHole is called

File: replica/backup.go (hunk 14 above)
Comment (vishnuitta) [reply replica/backup.go]: it seems - writeAt is also incrementing this..
so, do `d.UsedBlocks++` here, and we can do similar above check there during writes

**Comment by utkarshmani1997** [comment]
Revert to earlier approach, i.e, replica will be WO mode until Preload get completed. IO's will be served if quorum is met. (2 replica in RW mode out of 3)

File: controller/replicator.go [hunk 17]
```diff
@@ -505,3 +506,27 @@ func (r *replicator) GetCloneStatus(address string) (string, error) {
...
 	return status, nil
 }
+
+func (r *replicator) GetPreloadStatus(address string) error {
+	backend, ok := r.backends[address]
+	if !ok {
+		return fmt.Errorf("Cannot find backend %v", address)
+	}
+	for true {
+		status, err := backend.backend.GetPreloadStatus()
```
Comment (vishnuitta) [inline controller/replicator.go]: can we avoid using `backend` here?

File: replica/server.go [hunk 18]
```diff
@@ -376,6 +392,16 @@ func (s *Server) Close(signalMonitor bool) error {
 		return nil
 	}
 
+	if s.r.volume.preloadStatus != types.Done {
+		logrus.Warning("Already reading extents in background")
+		// verify if goroutine exited safely
+		for s.r.volume.preloadStatus == types.Started {
```
Comment (vishnuitta) [inline replica/server.go]: this `for` loop doesn't make sense

File: controller/control.go [hunk 19]
```diff
@@ -1035,3 +1035,14 @@ func (c *Controller) monitoring(address string, backend types.Backend) {
...
 }
+
+func (c *Controller) monitorPreload(address string) {
+	logrus.Infof("Get preload status of backend %s", address)
+	if err := c.backend.GetPreloadStatus(address); err != nil {
+		logrus.Errorf("Failed to monitor preload, error: %v", err)
+		return
+	}
+	c.startFrontend()
+	c.setReplicaModeNoLock(address, types.RW)
```
Comment (vishnuitta) [inline controller/control.go]: why 'NoLock'?

File: controller/replicator.go (hunk 17 above)
Comment (utkarshmani1997) [reply controller/replicator.go]: We can't, this is the way, we can get the PreloadStatus.

File: replica/server.go (hunk 18 above)
Comment (utkarshmani1997) [reply replica/server.go]: Why it doesn't make sense ?

File: controller/control.go (hunk 19 above)
Comment (utkarshmani1997) [reply controller/control.go]: There is no specific reason, since setReplicaModeNoLock() was getting used that's why used the same, because logic is no different that earlier.

Comment (vishnuitta) [reply controller/control.go]: its not about that current code. Won't we be needing lock here?

Comment (utkarshmani1997) [reply controller/control.go]: We are entering by taking lock, that's why we donot need another lock.

File: controller/control.go [hunk 20]
```diff
@@ -1035,3 +1038,14 @@ func (c *Controller) monitoring(address string, backend types.Backend) {
 	logrus.Infof("Monitoring stopped %v", address)
 	c.RemoveReplicaNoLock(address)
 }
+
+func (c *Controller) monitorPreload(address string) {
```
Comment (payes) [inline controller/control.go]: why are we not using any locks in this function.

File: controller/control.go [hunk 21]
```diff
@@ -706,7 +709,8 @@ getCloneStatus:
 	} else if status == "error" {
 		return fmt.Errorf("Replica clone status returned error %s", address)
 	}
-	c.setReplicaModeNoLock(address, types.RW)
+	c.setReplicaModeNoLock(address, types.WO)
```
Comment (payes) [inline controller/control.go]: I guess this is not required as the replica is already in WO mode.

Comment (utkarshmani1997) [reply controller/control.go]: This is required because we update the status to RO in UpdateVolStatus

File: controller/control.go (hunk 20 above)
Comment (utkarshmani1997) [reply controller/control.go]: have added in the recent commit

File: sync/sync.go [hunk 22]
```diff
@@ -357,9 +361,19 @@ Register:
...
+	if _, err := toClient.ReloadReplica(); err != nil {
+		logrus.Errorf("Error in reloadreplica %s", replicaAddress)
+		return err
+	}
+
+	// Running as a saparate goroutine such that if controller
+	// disconnected while preload, replica get notified immediately
+	// on MonitorChannel.
+	logrus.Info("Preload and verify rebuild")
+	go t.preloadAndVerifyRebuild(replicaAddress, toClient, s)
```
Comment (payes) [inline sync/sync.go]: What is the need to run it as a goroutine.

File: controller/control.go [hunk 23]
```diff
@@ -1035,3 +1038,16 @@ func (c *Controller) monitoring(address string, backend types.Backend) {
 	logrus.Infof("Monitoring stopped %v", address)
 	c.RemoveReplicaNoLock(address)
 }
+
+func (c *Controller) monitorPreload(address string) {
+	c.Lock()
```
Comment (payes) [inline controller/control.go]: Start controller and 3 replicas, immediately restart all 3 replicas.
Between the above 2 events, controller sends start signal to replica 1, and it is waiting for replica 1 to complete its preload with the lock and sleeping for 2 sec.
Replicas got restarted during this 2 sec, and the old IPs persist.
Controller again tries to get preload status on same IP after 2 sec, since controller doesn't know that the replica went down while it was sleeping. 
Preload status received by controller "Preload is not done yet on backend tcp://172.18.0.5:9502, current status: NA" Error at replica for the the above preloadStatus request "Fail to get revision counter: BUG: revision file wasn't initialized".
Controller tries it indefinitely while the GET request from replicas to controller are waiting for this lock.

File: sync/sync.go [hunk 24]
```diff
@@ -322,7 +322,11 @@ Register:
 	}
 	if action == "start" {
 		logrus.Infof("Received start from controller")
-		return t.client.Start(replicaAddress)
+		if err := t.client.Start(replicaAddress); err != nil {
+			return err
+		}
+		go s.Preload()
```
Comment (payes) [inline sync/sync.go]: Care also needs to be taken while calling s.Preload() without a lock. If s.r is closed while preload is going on, replica might show unexpected behavior.

File: sync/sync.go [hunk 25]
```diff
@@ -383,22 +397,25 @@ func (t *Task) checkAndResetFailedRebuild(address string, server *replica.Server
 	return nil
 }
 
-func (t *Task) reloadAndVerify(address string, repClient *replicaClient.ReplicaClient) error {
-	_, err := repClient.ReloadReplica()
-	if err != nil {
-		logrus.Errorf("Error in reloadreplica %s", address)
-		return err
+func (t *Task) preloadAndVerifyRebuild(address string, repClient *replicaClient.ReplicaClient, s *replica.Server) {
+	if err := s.Preload(); err != nil {
```
Comment (payes) [inline sync/sync.go]: Same as above comment

File: controller/control.go [hunk 26]
```diff
@@ -741,56 +737,29 @@ func (c *Controller) Start(addresses ...string) error {
...
-		if sendSignal == 1 {
-			logrus.Infof("sending add signal to quorum %v", regrep)
-			c.factory.SignalToAdd(regrep, "add")
-		}
-	}
-	logrus.Info("Update volume status")
-	c.UpdateVolStatus()
-
+	/*
+	   * This code has been commented since we start with only one replica currently
```
Comment (vishnuitta) [inline controller/control.go]: This commented code can be removed.. and, add a comment that this is moved to monitorPreload function

File: controller/control.go [hunk 27]
```diff
@@ -1035,3 +1006,68 @@ func (c *Controller) monitoring(address string, backend types.Backend) {
...
+		sendSignal = true
+		for _, tmprep := range c.replicas {
+			if tmprep.Address == "tcp://"+regrep+":9502" {
+				sendSignal = false
+				break
+			}
+		}
+		if sendSignal {
+			logrus.Infof("sending add signal to %v", regrep)
+			c.factory.SignalToAdd(regrep, "add")
```
Comment (vishnuitta) [inline controller/control.go]: This signaling after preload is important. In previous versions, this was missed, and would have caused some data issue.

File: types/types.go [hunk 28]
```diff
@@ -134,6 +140,7 @@ type Frontend interface {
 }
 
 type DataProcessor interface {
+	Close(signalMonitor bool) error
```
Comment (vishnuitta) [inline types/types.go]: I don't think this change is required now.

File: replica/replica.go [hunk 29]
```diff
@@ -173,7 +173,7 @@ func construct(readonly bool, size, sectorSize int64, dir, head string, backingF
 		r.info.BackingFileName = backingFile.Name
 	}
 	r.volume.sectorSize = defaultSectorSize
-
+	r.volume.preloadStatus = types.None
```
Comment (vishnuitta) [inline replica/replica.go]: lets add a log at all the status change of preload to help in debugging

File: replica/replica.go [hunk 30]
```diff
@@ -217,11 +217,20 @@ func construct(readonly bool, size, sectorSize int64, dir, head string, backingF
 	r.insertBackingFile()
 	r.ReplicaType = replicaType
 
+	return r, r.writeVolumeMetaData(true, r.info.Rebuilding)
```
Comment (vishnuitta) [inline replica/replica.go]: need to look at other callers of construct (clone? revert? etc) so that Preload is calling after this call

File: replica/server.go [hunk 31]
```diff
@@ -86,6 +88,9 @@ func (s *Server) Create(size int64) error {
 	state, _ := s.Status()
 
 	if state != Initial {
+		r := &Replica{}
```
Comment (vishnuitta) [inline replica/server.go]: why this change?

**Comment by utkarshmani1997** [comment]
Closing this PR, since it doesn't cover clone and revert cases. Raised new PR #181 for the same. This PR can be opened if need arise in future.
